integration/internal/process_test.go

```diff
@@ -28,45 +37,143 @@ var _ = Describe("Start method", func() {
 		Consistently(processState.Session.ExitCode).Should(BeNumerically("==", -1))
 	})
 
-	Context("when process takes too long to start", func() {
-		It("returns a timeout error", func() {
-			processState := &ProcessState{}
-			processState.Path = "bash"
-			processState.Args = simpleBashScript
-			processState.StartTimeout = 200 * time.Millisecond
-			processState.StartMessage = "loop 5000"
-
-			err := processState.Start(nil, nil)
-			Expect(err).To(MatchError(ContainSubstring("timeout")))
+	Context("when a health check endpoint is provided", func() {
+		var server *ghttp.Server
+		BeforeEach(func() {
+			server = ghttp.NewServer()
 
-			Eventually(processState.Session.ExitCode).Should(Equal(143))
 		})
-	})
+		AfterEach(func() {
+			server.Close()
+		})
 
-	Context("when the command cannot be started", func() {
-		var (
-			processState *ProcessState
-		)
-		BeforeEach(func() {
-			processState = &ProcessState{}
-			processState.Path = "/nonexistent"
+		Context("when the healthcheck returns ok", func() {
+			BeforeEach(func() {
+				server.RouteToHandler("GET", "/healthz", ghttp.RespondWith(http.StatusOK, ""))
```

Comment:
This is a pattern we are just used to use over here. I can see that it might look strange and overly complicated / too exhaustive / ... sometimes. Just to give you a bit of context, why we structured the tests as they are:

- We have the different `Context`s which tell a different (sub)story of what we want to test in this particular case
- In each `Context` we create the state as we described it in the title of the `Context`, mostly with `{Just,}BeforeEach` or so
- In the `It` we do the actual test, do what we have described in the title of the `It`, with the assumption that the context is already prepared as described

So it is not so much about sharing setup code for multiple tests (at least in this example), for us it is more about the story telling (and arguably(?) nicely readable ginkgo output maybe).

If I'd pull the `server.RouteToHandler("GET", "/healthz", ...)
` from the `Context`'s `BeforeEach` into the `It`, I'd feel an urge to change the title of the `It` to something like
> hits an endpoint which always returns OK, and successfully starts

Does my long-winded explanation make any sense to you?

Reply:
It's still not clear to me why some code belongs in BeforeEach and other code does not, and I'm concerned that the distinction is arbitrary.  I think separating out code in arbitrary ways imposes costs to readability and maintainability, because anything that's arbitrary is difficult to communicate or have common standards for.  I'm not going to force you to change, I'm just providing my perspective.